scylla-cql/src/types/deserialize/row.rs

```diff
@@ -122,6 +123,96 @@ impl<'frame> DeserializeRow<'frame> for ColumnIterator<'frame> {
+        where
+            $($Ti: DeserializeValue<'frame>),*
+        {
+            fn type_check(specs: &[ColumnSpec]) -> Result<(), TypeCheckError> {
+                const TUPLE_LEN: usize = (&[$($idx),*] as &[i32]).len();
+
+                if let [$($idf),*] = &specs {
+                    $(
+                        <$Ti as DeserializeValue<'frame>>::type_check(&$idf.typ)
+                            .map_err(|err| mk_typck_err::<Self>(BuiltinTypeCheckErrorKind::ColumnTypeMismatch {
+                                column_index: $idx,
+                                column_name: specs[$idx].name.clone(),
+                                err
+                            }))?;
+                    )*
+                    Ok(())
+                } else {
+                    Err(mk_typck_err::<Self>(BuiltinTypeCheckErrorKind::WrongColumnCount {
+                        rust_cols: TUPLE_LEN, cql_cols: specs.len()
+                    }))
+                }
+            }
+
+            fn deserialize(mut row: ColumnIterator<'frame>) -> Result<Self, DeserializationError> {
+                const TUPLE_LEN: usize = (&[$($idx),*] as &[i32]).len();
+
+                let ret = (
+                    $({
+                        let column = row.next().unwrap_or_else(|| unreachable!(
+                            "Typecheck should have prevented this scenario! Column count mismatch: rust type {}, cql row {}",
+                            TUPLE_LEN,
+                            $idx
+                        )).map_err(deser_error_replace_rust_name::<Self>)?;
+
+                        <$Ti as DeserializeValue<'frame>>::deserialize(&column.spec.typ, column.slice)
+                            .map_err(|err| mk_deser_err::<Self>(BuiltinDeserializationErrorKind::ColumnDeserializationFailed {
+                                column_index: column.index,
+                                column_name: column.spec.name.clone(),
+                                err,
+                            }))?
+                    },)*
+                );
+                assert!(
+                    row.next().is_none(),
+                    "Typecheck should have prevented this scenario! Column count mismatch: rust type {}, cql row is bigger",
+                    TUPLE_LEN,
+                );
+                Ok(ret)
+            }
+        }
+    }
+}
+
+macro_rules! impl_tuple_multiple {
+    (;< $($T_PREF:ident,)* >;;< $($idx_pref:literal,)* >;;< $($idf_pref:ident,)* >) => {
+        impl_tuple!(;;);
+    };
+    ($TN:ident $(,$Ti:ident)*; < $($T_PREF:ident,)* > ;  $idx_n:literal $(,$idx:literal)*; < $($idx_pref:literal,)* >; $idf_n:ident $(,$idf:ident)* ; <$($idf_pref:ident,)*>) => {
+        impl_tuple!($($T_PREF,)* $TN ; $($idx_pref, )* $idx_n; $($idf_pref, )* $idf_n);
+        impl_tuple_multiple!($($Ti),* ; < $($T_PREF,)* $TN, > ; $($idx),*; < $($idx_pref, )* $idx_n,  >; $($idf),*; < $($idf_pref, )* $idf_n, >);
+    }
+}
```

Comment:
Again, some comment would be nice

Reply:
What kind of a comment would you like here?
What comes to my mind is "Macro to implement tuple deserialization for various sizes of tuple", which is not much helpful.